File: templates/python/app/workflows/generator.py
from moose_lib import Task, TaskConfig, Workflow, WorkflowConfig, OlapTable, InsertOptions, Key, TaskContext
from pydantic import BaseModel
from datetime import datetime
from faker import Faker
from app.ingest.models import Foo, Baz
import requests

class FooWorkflow(BaseModel):
    id: Key[str]
    success: bool
    message: str

workflow_table = OlapTable[FooWorkflow]("foo_workflow")

def run_task(ctx: TaskContext[None]) -> None:
    fake = Faker()
    for i in range(1000):
        # Prepare request data
        foo = Foo(
            primary_key=fake.uuid4(),
            timestamp=fake.date_time_between(start_date='-1y', end_date='now').timestamp(),
            baz=fake.random_element(Baz),
            optional_text=fake.text() if fake.boolean() else None
        )
 
        # POST record to Moose Ingest API
        req = requests.post(
            "http://localhost:4000/ingest/Foo",
            data=foo.model_dump_json().encode('utf-8'),
            headers={'Content-Type': 'application/json'}
        )

        if req.status_code == 200:
            workflow_table.insert([{"id": "1", "success": True, "message": f"Inserted Foo with primary key: {foo.primary_key}"}])
        else:
            workflow_table.insert([{"id": "1", "success": False, "message": f"Failed to insert Foo with error: {req.status_code}"}])

ingest_task = Task[None, None](
    name="task",
    config=TaskConfig(run=run_task)
)

ingest_workflow = Workflow(
    name="generator",
    config=WorkflowConfig(starting_task=ingest_task, retries=3, timeout="30s") ## add schedule="@every 5s" if you want to run it automatically on a schedule
)

File: templates/typescript/app/workflows/generator.ts
import { Task, Workflow, OlapTable, Key } from "@514labs/moose-lib";
import { Foo } from "../ingest/models";
import { faker } from "@faker-js/faker";

// Data model for OLAP Table
interface FooWorkflow {
  id: Key<string>;
  success: boolean;
  message: string;
}

// Create OLAP Table
const workflowTable = new OlapTable<FooWorkflow>("FooWorkflow");

export const ingest = new Task<null, void>("ingest", {
  run: async () => {
    for (let i = 0; i < 1000; i++) {
      const foo: Foo = {
        primaryKey: faker.string.uuid(),
        timestamp: faker.date.recent({ days: 365 }).getTime(),
        optionalText: Math.random() < 0.5 ? faker.lorem.text() : undefined,
      };

      try {
        const response = await fetch("http://localhost:4000/ingest/Foo", {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(foo),
        });

        if (!response.ok) {
          console.log(
            `Failed to ingest record ${i}: ${response.status} ${response.statusText}`,
          );
          // Insert ingestion result into OLAP table
          workflowTable.insert([
            { id: "1", success: false, message: response.statusText },
          ]);
        }
      } catch (error) {
        console.log(`Error ingesting record ${i}: ${error}`);
        workflowTable.insert([
          { id: "1", success: false, message: error.message },
        ]);
      }

      // Add a small delay to avoid overwhelming the server
      if (i % 100 === 0) {
        console.log(`Ingested ${i} records...`);
        workflowTable.insert([
          { id: "1", success: true, message: `Ingested ${i} records` },
        ]);
        await new Promise((resolve) => setTimeout(resolve, 100));
      }
    }
  },
  retries: 3,
  timeout: "30s",
});

export const workflow = new Workflow("generator", {
  startingTask: ingest,
  retries: 3,
  timeout: "30s",
  // schedule: "@every 5s",
});
